File: test-engine.mjs
// Mirrors scripts/test_engine.py so the browser engine is held to the same bar.
import { Game, botAction, HAND } from './engine.js';

const failures = [];
const check = (cond, msg) => { if (!cond) failures.push(msg); };
const total = g => g.deck.length + g.discard.length
  + g.seats.reduce((n, s) => n + s.hand.length, 0);

function playOne(mode, matchType, seed) {
  const g = new Game({ mode, matchType, seed });
  const start = total(g);
  let steps = 0;
  while (!g.over && steps < 4000) {
    steps++;
    const actors = g.seats.map(s => s.index).filter(i => g.legalActions(i).length);
    check(actors.length <= 1,
      `${mode}/${matchType}: ${actors.length} seats could act at once in ${g.phase}`);
    if (!actors.length) { failures.push(`${mode}/${matchType}: deadlock in ${g.phase}`); break; }
    const seat = actors[0];

    const v = g.view(seat);
    const mine = new Set(v.hand.map(c => c.id));
    for (const other of g.seats) {
      if (other.index === seat) continue;
      check(!other.hand.some(c => mine.has(c.id)),
        'a seat view contained another seat card id');
    }
    check(v.seats.every(s => s.hand === undefined), 'seat summaries leaked hands');

    let rejected = false;
    try { g.apply(seat, { type: 'play', card_id: 'nope', face: 'PASS' }); }
    catch { rejected = true; }
    check(rejected, 'engine accepted an action for a card not in hand');

    g.apply(seat, botAction(g, seat, seat % 2 ? 'PATIENT' : 'SHOOTER'));
    check(total(g) === start, `card count drifted: ${total(g)} vs ${start}`);
    for (const s of g.seats) {
      check(s.hand.length <= HAND + 3, `hand grew to ${s.hand.length} in ${mode}`);
    }
  }
  check(g.over, `${mode}/${matchType}: did not finish in ${steps} steps`);
  return { g, steps };
}

const matches = Number(process.argv[2] || 250);
const tally = {};
const lengths = [];
for (let i = 0; i < matches; i++) {
  for (const mode of ['LUCK', 'STRATEGY']) {
    for (const mt of ['ONE_V_ONE', 'TWO_V_TWO']) {
      const { g, steps } = playOne(mode, mt, i * 7919 + mode.length * 31 + mt.length);
      lengths.push(steps);
      const reason = [...g.log].reverse().find(e => e.kind === 'match_over')?.reason || '?';
      const k = `${mode}/${mt}/${reason}`;
      tally[k] = (tally[k] || 0) + 1;
    }
  }
}
lengths.sort((a, b) => a - b);
console.log(`\nplayed ${matches * 4} matches in the browser engine`);
console.log(`median actions per match: ${lengths[Math.floor(lengths.length / 2)]}`);
for (const k of Object.keys(tally).sort()) console.log(`  ${k.padEnd(34)}${tally[k]}`);
if (failures.length) {
  console.log(`\n${failures.length} FAILURES`);
  [...new Set(failures)].slice(0, 12).forEach(f => console.log(`  • ${f}`));
  process.exit(1);
}
console.log('\nall invariants held: no leaks, no illegal actions, no lost cards, every match finished.\n');

File: test_engine.py
#!/usr/bin/env python3
"""
Engine tests. Plays full matches through the same API the server will use, and
checks the invariants that matter for a networked card game:

  1. No illegal action is ever accepted.
  2. A seat's view never contains another seat's cards.
  3. Hands return to 4 whenever a turn boundary is reached.
  4. Total cards in the system stays constant — no card is duplicated or lost.
  5. Every match terminates.
  6. Only the seat named by the phase has legal actions.

Run: python scripts/test_engine.py [--matches 300]
"""

import argparse
import random
from collections import Counter

from engine import HAND, Game, bot_action

FAIL = []


def check(cond, msg):
    if not cond:
        FAIL.append(msg)


def total_cards(g):
    return len(g.deck) + len(g.discard) + sum(len(s.hand) for s in g.seats)


def play_one(mode, match_type, seed):
    g = Game(mode=mode, match_type=match_type, seed=seed)
    start_total = total_cards(g)
    steps = 0

    while not g.over and steps < 4000:
        steps += 1

        actors = [s.index for s in g.seats if g.legal_actions(s.index)]
        check(len(actors) <= 1,
              f"{mode}/{match_type}: {len(actors)} seats could act at once "
              f"in phase {g.phase}")
        if not actors:
            FAIL.append(f"{mode}/{match_type}: deadlock in phase {g.phase}")
            break
        seat = actors[0]

        # views must not leak
        v = g.view(seat)
        mine = {c["id"] for c in v["hand"]}
        for other in g.seats:
            if other.index == seat:
                continue
            check(not (mine & {c.id for c in other.hand}),
                  "a seat's view contained another seat's card ids")
        check(all(k != "hand" for k in v["seats"][0].keys()),
              "seat summaries leaked hands")

        # an action outside the legal list must be refused
        try:
            g.apply(seat, {"type": "play", "card_id": "does-not-exist",
                           "face": "PASS"})
            FAIL.append("engine accepted an action for a card not in hand")
        except ValueError:
            pass

        g.apply(seat, bot_action(g, seat,
                                "PATIENT" if seat % 2 else "SHOOTER"))

        check(total_cards(g) == start_total,
              f"card count drifted: {total_cards(g)} vs {start_total}")
        for s in g.seats:
            check(len(s.hand) <= HAND + 3,
                  f"hand grew to {len(s.hand)} in {mode}")

    check(g.over, f"{mode}/{match_type}: match did not finish in {steps} steps")
    return g, steps


def main(matches):
    random.seed(1)
    stats = Counter()
    lengths = []
    for i in range(matches):
        for mode in ("LUCK", "STRATEGY"):
            for mt in ("ONE_V_ONE", "TWO_V_TWO"):
                g, steps = play_one(mode, mt, seed=i * 97 + hash(mode + mt) % 1000)
                lengths.append(steps)
                if g.over:
                    reason = next((e["reason"] for e in reversed(g.log)
                                   if e["kind"] == "match_over"), "?")
                    stats[f"{mode}/{mt}/{reason}"] += 1

    print(f"\nplayed {matches * 4} matches "
          f"(2 modes x 2 formats x {matches} seeds)")
    print(f"median actions per match: {sorted(lengths)[len(lengths)//2]}")
    for k, v in sorted(stats.items()):
        print(f"  {k:<34}{v}")

    if FAIL:
        print(f"\n{len(FAIL)} FAILURES")
        for f in sorted(set(FAIL))[:12]:
            print(f"  • {f}")
        return 1
    print("\nall invariants held: no leaks, no illegal actions, "
          "no lost cards, every match finished.\n")
    return 0


if __name__ == "__main__":
    ap = argparse.ArgumentParser()
    ap.add_argument("--matches", type=int, default=300)
    a = ap.parse_args()
    raise SystemExit(main(a.matches))
